File: src/main/python/stack.py
class ListItem:
	next = None
	val = 0

class Stack:
	top = None

	def push(self, val):
		elt = ListItem()
		elt.val = val
		elt.next = self.top
		self.top = elt

	def pop(self):
		if self.top != None:
			tmp = self.top
			self.top = tmp.next
			return tmp
		else:
			return None


stack = Stack()
stack.push(1)
print stack.pop().val
print stack.pop()

stack.push(2)
stack.push(3)

print stack.pop().val
print stack.pop().val

File: src/main/java/org/sherman/interview/stack/Stack.java
package org.sherman.interview.stack;

import com.google.common.base.Joiner;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toList;

/**
 * @author Denis Gabaydulin
 * @since 04.12.16
 */
public class Stack<T> {
    private Object[] elts;
    private int size;

    public Stack(int size) {
        elts = new Object[size];
    }

    public void push(@NotNull T elt) {
        if (elts.length == size) {
            resize();
        }

        elts[size++] = elt;
    }

    @NotNull
    public T pop() {
        if (size == 0) {
            throw new IllegalStateException("Stack is empty!");
        }
        return (T) elts[--size];
    }

    public T peek() {
        if (size == 0) {
            return null;
        }
        return (T) elts[size - 1];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public String toString() {
        return "[" + (size > 0 ? Joiner.on(";").join(getNonNullElts()) : "") + "]";
    }

    private List<Object> getNonNullElts() {
        return Arrays.stream(elts)
                .filter(Objects::nonNull)
                .collect(toList());
    }

    private void resize() {
        Object[] newElts = new Object[size * 2];
        System.arraycopy(elts, 0, newElts, 0, size);
    }
}
